[JSC] Fix SIMD in new BBQ

Constellation opened the pull request:
#### 672628fb19da7b1b93ab08eaa288fdb3ef5897d9
<pre>
[JSC] Fix SIMD in new BBQ
<a href="https://bugs.webkit.org/show_bug.cgi?id=252736">https://bugs.webkit.org/show_bug.cgi?id=252736</a>
rdar://105769837

Reviewed by Mark Lam.

Fix SIMD ops in new BBQ.

1. Some ops uses possibly-overwritten FPR registers. Fixed with scratch registers.
2. Clean up AirLowerMacros&apos;s generation.
3. SIMD splat can take Int constants. Thus we should check and load constants.

* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDSplat):
(JSC::Wasm::BBQJIT::addSIMDI_V):
(JSC::Wasm::BBQJIT::addSIMDV_V):
(JSC::Wasm::BBQJIT::addSIMDRelOp):
(JSC::Wasm::BBQJIT::emitVectorMul):

Canonical link: <a href="https://commits.webkit.org/260700@main">https://commits.webkit.org/260700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15674712af937d24dbea2ccd469baaf7d2b26fab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109111 "Failed to checkout and rebase branch from PR 10507") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18193 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9495 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101351 "Built successfully") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/114869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10975 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/9061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11712 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17080 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/106601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13320 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/106601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4037 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -7204,15 +7215,17 @@ class BBQJIT {
-                m_jit.vectorXor(info, valueLocation.asFPR(), m_scratchFPR, resultLocation.asFPR());
+            // Equality is always a single instruction for this lane, so it doesn't matter that our scratch register is the same as the operands.
+            ScratchScope<0, 1> scratches(*this, valueLocation, resultLocation);
+            m_jit.compareIntegerVector(RelationalCondition::Equal, SIMDInfo { SIMDLane::i32x4, SIMDSignMode::None }, m_scratchFPR, m_scratchFPR, m_scratchFPR, scratches.fpr(0));
```
Has this actually caused any bugs? The scratch parameter to `compareIntegerVector` can be clobbered generally, but for `RelationalCondition::Equal` specifically, we shouldn't use it (hence the comment I left above). So I figured we might as well avoid an extra unnecessary register.

Constellation replied:
I think it is unsafe that passing the same register as a scratchFPR to MacroAssembler. MacroAssembler is thinking that scratchFPR is separate from the other registers and do not assume that the above is safe.

MenloDorian replied:
Let's remove that comment since it is incorrect.

Constellation replied:
Removed

ddegazio replied:
I guess my thinking is, I can't imagine us needing to change the macro assembler implementation, so it's probably not too problematic to assume this behavior (since we know the architecture and instruction we are generating exactly). Whereas reserving an extra register could cause us to spill unnecessarily. But I suppose this change is fine.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -7349,11 +7370,12 @@ class BBQJIT {
             m_jit.vectorExtractLaneInt64(TrustedImm32(0), right.asFPR(), m_dataScratchGPR);
             m_jit.mul64(m_scratchGPR, m_dataScratchGPR, m_scratchGPR);
-            m_jit.vectorReplaceLane(SIMDLane::i64x2, TrustedImm32(0), m_scratchGPR, result.asFPR());
+            m_jit.vectorReplaceLane(SIMDLane::i64x2, TrustedImm32(0), m_scratchGPR, m_scratchFPR);
```
Same as above, does this actually cause issues? Even though we write to `result.asFPR()` prior to the use of `left/right.asFPR()` below (which may alias), my expectation is that the lanes are distinct, and for each lane we obey the ordering that we write the result only after reading each input.

(For this one I'm mostly just curious, I still think the change looks good since it's less fragile and doesn't require additional scratches)

Constellation replied:
Yes, `JSTests/wasm/v8/simd-i64x2-mul.js` is failing due to this.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Fix SIMD in new BBQ
https://bugs.webkit.org/show_bug.cgi?id=252736
rdar://105769837

Reviewed by Mark Lam.

Fix SIMD ops in new BBQ.

1. Some ops uses possibly-overwritten FPR registers. Fixed with scratch registers.
2. Clean up AirLowerMacros's generation.
3. SIMD splat can take Int constants. Thus we should check and load constants.

* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDSplat):
(JSC::Wasm::BBQJIT::addSIMDI_V):
(JSC::Wasm::BBQJIT::addSIMDV_V):
(JSC::Wasm::BBQJIT::addSIMDRelOp):
(JSC::Wasm::BBQJIT::emitVectorMul):

Canonical link: https://commits.webkit.org/260700@main
